core/services/ocr2/plugins/s4/plugin.go

```diff
@@ -0,0 +1,230 @@
+package s4
+
+import (
+	"context"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
+	"github.com/smartcontractkit/chainlink/v2/core/services/s4"
+
+	"github.com/pkg/errors"
+	"github.com/smartcontractkit/libocr/offchainreporting2/types"
+)
+
+type plugin struct {
+	logger       logger.Logger
+	config       *PluginConfig
+	orm          s4.ORM
+	addressRange *s4.AddressRange
+}
+
+type key struct {
+	address string
+	slot    uint
+}
+
+var _ types.ReportingPlugin = (*plugin)(nil)
+
+func NewReportingPlugin(logger logger.Logger, config *PluginConfig, orm s4.ORM) (types.ReportingPlugin, error) {
+	if config.MaxObservationEntries == 0 {
+		return nil, errors.New("max number of observation entries cannot be zero")
+	}
+	if config.MaxReportEntries == 0 {
+		return nil, errors.New("max number of report entries cannot be zero")
+	}
+
+	addressRange, err := s4.NewInitialAddressRangeForIntervals(config.NSnapshotShards)
+	if err != nil {
+		return nil, err
+	}
+
+	return &plugin{
+		logger:       logger.Named("OCR2-S4").With("product", config.Product),
+		config:       config,
+		orm:          orm,
+		addressRange: addressRange,
+	}, nil
+}
+
+func (c *plugin) Query(ctx context.Context, _ types.ReportTimestamp) (types.Query, error) {
+	promReportingPluginQuery.WithLabelValues(c.config.Product).Inc()
+
+	ormVersions, err := c.orm.GetVersions(c.addressRange, pg.WithParentCtx(ctx))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to GetVersions in Query()")
+	}
+
+	versions := make([]*VersionRow, len(ormVersions))
+	for i, v := range ormVersions {
+		versions[i] = &VersionRow{
+			Address: v.Address.Hex(),
+			Slotid:  uint32(v.SlotId),
+			Version: v.Version,
+		}
+	}
+
+	queryBytes, err := MarshalQuery(versions)
+	if err != nil {
+		return nil, err
+	}
+
+	promReportingPluginsQueryRowsCount.WithLabelValues(c.config.Product).Set(float64(len(versions)))
+	promReportingPluginsQueryByteSize.WithLabelValues(c.config.Product).Set(float64(len(queryBytes)))
+
+	c.addressRange.Advance()
+
+	return queryBytes, err
+}
+
+func (c *plugin) Observation(ctx context.Context, _ types.ReportTimestamp, query types.Query) (types.Observation, error) {
+	promReportingPluginObservation.WithLabelValues(c.config.Product).Inc()
+
+	if err := c.orm.DeleteExpired(pg.WithParentCtx(ctx)); err != nil {
+		return nil, errors.Wrap(err, "failed to DeleteExpired in Observation()")
+	}
+
+	observationRows := make([]*s4.Row, 0)
```

Comment:
This name is pretty confusing. How about "leaderUpdateRows" or "queryResponseRows" ... not great names either but let's figure out something that matches the content.

Reply:
Renamed to `queryRows` at least. No idea for a better naming here.